### PR TITLE
Autonaming

### PR DESCRIPTION
**Pull request by mikhailshilkov**
Resolves https://github.com/pulumi/pulumi-google-native/issues/93 for the majority of resources. I will open an issue to track the remaining ones.

This PR applies auto-naming if
- There is an input property called `name`, and
- Its description doesn't contain a forward slash, indicating a simple name structure, or
- We can parse the description to extract the pattern for the name property/ (like `projects/*/locations/*/functions/function_name`)

My estimate is that it covers 90+% of resources. The others may need manual overrides, either in the regex or as an exclusion list.

### CODE REVIEW

**Comment by lblackstone** [comment]
The implementation looks fine, but it does seem like this could be confusing for users. It's easy to explain if auto-naming is either supported or not, but seems like a pain given the number of resources that don't support it yet.

I wonder if it might be better to merge the provider changes needed to support it, but hold off on SDK changes until we can figure out a more generalized solution?

File: provider/pkg/gen/schema.go
Comment (viveklak) [inline provider/pkg/gen/schema.go:443]: Just a thought - should we add a blurb to the description for the resource (or name property) to say that it supports autonaming?

Comment (viveklak) [reply provider/pkg/gen/schema.go:443]: Or we could flip it to mark out resources that don't support autonaming yet. Just a compromise while we figure out how to improve overall coverage. I think getting some autonaming support through - even if partial is a win. I see a decent number of highly used resources already on the list...

**Comment by mikhailshilkov** [comment]
@lblackstone @viveklak Please take a look at my last commit. It's not ready to merge but it's a direction that I'd love to validate with you.

Basically, it parses descriptions with a regex to extract the format of the name property. I haven't spent much time to make this regex universal but it does capture a bunch of cases. Again, we won't probably get to 100% but may eventually get to 99% with this approach. I'm fine with adding a blurb to the description of the others.

The same approach can be used for #56 I suppose.

If I could get your blessing today, I'll try to complete this PR during your long weekend. Thank you!

**Comment by infin8x** [comment]
Sorry for the delayed response. I was initially with @lblackstone that this could get confusing, but if we're able to get the autonaming-supported set of resources above, say, 85-90% with your latest commit, I definitely think it's worth getting in.

**Comment by lblackstone** [comment]
The updated approach LGTM. I think it's worth taking if we can cover a majority of cases and include docs that make it obvious where it's not yet supported.

**Comment by mikhailshilkov** [comment]
@lblackstone @viveklak this is ready for review

File: provider/pkg/gen/utilities.go
Comment (viveklak) [inline provider/pkg/gen/utilities.go:34]: Would it be a little more readable to break the `or`ed regex up and apply separately?  Do we intend to capture all these groups or just the ones with the paths is sufficient? Could we use a named group for that, e.g. https://golang.org/pkg/regexp/#Regexp.SubexpNames.

Comment (viveklak) [reply provider/pkg/gen/utilities.go:34]: In general would be good to annotate regexes with an example or two of the type of strings they are supposed to match.

**Review by viveklak — COMMENTED** [review]
Should we also add an example for a resource that we derive autonaming pattern from the description for?

**Comment by mikhailshilkov** [comment]
> Should we also add an example for a resource that we derive autonaming pattern from the description for?

@viveklak  Not sure what exactly you mean here. Most of the test cases in unit tests are examples of real resources. Do you mean something else?

File: provider/pkg/gen/utilities.go
Comment (mikhailshilkov) [reply provider/pkg/gen/utilities.go]: I split the expression into two and added an example to each. Now there's just one subgroup that we care about, so it should be fine without a name.

File: provider/pkg/gen/schema.go
Comment (lblackstone) [inline provider/pkg/gen/schema.go:440]: I think it might be clearer to include the HasSuffix condition as an OR on the first branch, and then use a normal `else` branch. I'm also not quite clear on why IamPolicy is special, so you might consider including that condition in a named function.

**Review by lblackstone — APPROVED** [review]

File: provider/pkg/gen/schema.go
Comment (mikhailshilkov) [reply provider/pkg/gen/schema.go]: Refactored to a named variable and an explicit if statement on top

**Pull request merged**
